Userland/Services/DeviceMapper/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,241 @@
+    for (auto& matcher : s_matchers) {
+        if (matcher.major_number == major_number
+            && block_device == matcher.block_device
+            && is_in_minor_number_range(matcher, minor_number))
+            return matcher;
+    }
+    return {};
+}
+
+static bool is_in_family_minor_number_range(DeviceNodeFamily const& family, MinorNumber minor_number)
+{
+    return family.base_minor_number() <= minor_number && static_cast<MinorNumber>(family.base_minor_number().value() + family.devices_symbol_suffix_allocation_map().size()) >= minor_number;
+}
+
+Optional<DeviceNodeFamily&> DeviceEventLoop::find_device_node_family(bool block_device, MajorNumber major_number, MinorNumber minor_number) const
+{
+    DeviceNodeFamily::Type type = block_device ? DeviceNodeFamily::Type::CharacterDevice : DeviceNodeFamily::Type::CharacterDevice;
+    for (auto const& family : m_device_node_families) {
+        if (family->major_number() == major_number && type == family->type() && is_in_family_minor_number_range(*family, minor_number))
+            return *family.ptr();
+    }
+    return {};
+}
+
+ErrorOr<NonnullRefPtr<DeviceNodeFamily>> DeviceEventLoop::find_or_register_new_device_node_family(DeviceNodeMatch const& match, bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    if (auto possible_family = find_device_node_family(block_device, major_number, minor_number); possible_family.has_value())
+        return possible_family.release_value();
+    unsigned allocation_map_size = 1024;
+    if (match.minor_number_allocation_type == MinorNumberAllocationType::SequentialLimited)
+        allocation_map_size = match.minor_number_range_size;
+    auto bitmap = TRY(Bitmap::create(allocation_map_size, false));
+    auto node = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) DeviceNodeFamily(move(bitmap),
+        match.family_type_literal,
+        block_device ? DeviceNodeFamily::Type::BlockDevice : DeviceNodeFamily::Type::CharacterDevice,
+        major_number,
+        minor_number)));
+    TRY(m_device_node_families.try_append(node));
+
+    return node;
+}
+
+static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
+{
+    auto base_string = TRY(String::from_utf8(""sv));
+    do {
+        base_string = TRY(String::formatted("{:c}{}", 'a' + (allocation_index % 26), base_string));
+        allocation_index = allocation_index / 26;
+    } while (allocation_index > 0);
+    return base_string;
+}
+
+static ErrorOr<String> build_suffix_with_numbers(size_t allocation_index)
+{
+    return String::number(allocation_index);
+}
+
+static ErrorOr<void> prepare_permissions_after_populating_devtmpfs(StringView path, DeviceEventLoop::DeviceNodeMatch const& match)
+{
+    if (match.permission_group.is_null())
+        return {};
+    auto group = TRY(Core::System::getgrnam(match.permission_group));
+    VERIFY(group.has_value());
+    TRY(Core::System::chown(path, 0, group.value().gr_gid));
+    TRY(Core::System::endgrent());
```

Comment:
Apparently our LibC's `getgrnam` performs an automatic `endgrent` but `Core::System::getgrnam` does not; either we should align that behavior or this `endgrent` should take into account the possibility that `chown` fails and the group database isn't closed as a result of that.